### PR TITLE
Update Custom Formmatter

### LINKED ISSUE #16486: Update configuring for 3.1
In the "How to configure MVC to use a custom formatter", it sets the formatters in the lambda for services.AddMvc(). This was deprecated in 3.0. How do I configure ASP.NET to use my custom formatters in 3.0?

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: e2207f37-0cf3-3ab9-add8-c50c36409b53
* Version Independent ID: df89a2fd-501e-c917-2962-e4d8d124e852
* Content: [Custom formatters in ASP.NET Core Web API](https://docs.microsoft.com/en-us/aspnet/core/web-api/advanced/custom-formatters?view=aspnetcore-3.1)
* Content Source: [aspnetcore/web-api/advanced/custom-formatters.md](https://github.com/aspnet/AspNetCore.Docs/blob/master/aspnetcore/web-api/advanced/custom-formatters.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-webapi**
* GitHub Login: @Rick-Anderson
* Microsoft Alias: **riande**

### PR DESCRIPTION
**Pull request by Rick-Anderson**
Fixes  #16486
[Internal review URL](https://review.docs.microsoft.com/en-us/aspnet/core/web-api/advanced/custom-formatters?view=aspnetcore-3.1&branch=pr-en-us-18795)

### CODE REVIEW

**Comment by Rick-Anderson** [comment]
@serpent5  the 2.2 and the new 3.1 version work fine with GET, but on `POST` the model is null.

I changed 
Content-Type header "text/vcard" 
to
Content-Type header `text/vcard`
As Content-Type header `"text/vcard"` yields unsupported media type in Postman

cc @tdykstra

**Comment by Rick-Anderson** [comment]
@tdykstra  found the problem, @guardrex  guardRegEx accidently put this line in the sample

```
no-loc: [Blazor, "Identity", "Let's Encrypt", Razor, SignalR]
```

@serpent5  hand off, I need to remove steve's repository that 99% of customers don't like.

**Comment by guardrex** [comment]
Yes, the find-'n-**_rexplace_** was keyed on inserting the no-loc line at "uid:" with case insensitivity, so it took down the uppercase in the process.

**Comment by Rick-Anderson** [comment]
```                 
   // TODO review, UID is internal only, it's not Vcard data
                    //var idLine = await ReadLineAsync("uid:", reader, context);
                    //contact.ID = idLine.Substring(4);

// Read but never used
                    await ReadLineAsync("FN:", reader, context, logger);

```

**Comment by Rick-Anderson** [comment]
@serpent5  please review,, no hurry.

**Comment by serpent5** [comment]
@Rick-Anderson I will look at this within the next few days.

File: aspnetcore/web-api/advanced/custom-formatters.md [hunk 1]
```diff
@@ -9,112 +9,135 @@ uid: web-api/advanced/custom-formatters
...
 
-This article shows how to add support for additional formats by creating custom formatters. For an example of a custom input formatter for plain text, see [TextPlainInputFormatter](https://github.com/aspnet/Entropy/blob/master/samples/Mvc.Formatters/TextPlainInputFormatter.cs) on GitHub.
+This article shows how to add support for additional formats by creating custom formatters. For an example of a custom plain text input formatter, see [TextPlainInputFormatter](https://github.com/aspnet/Entropy/blob/master/samples/Mvc.Formatters/TextPlainInputFormatter.cs) on GitHub.
 
 [View or download sample code](https://github.com/dotnet/AspNetCore.Docs/tree/master/aspnetcore/web-api/advanced/custom-formatters/sample) ([how to download](xref:index#how-to-download-a-sample))
 
 ## When to use custom formatters
 
-Use a custom formatter when you want the [content negotiation](xref:web-api/advanced/formatting#content-negotiation) process to support a content type that isn't supported by the built-in formatters.
+Use a custom formatter for [content negotiation](xref:web-api/advanced/formatting#content-negotiation) types that aren't supported by the built-in formatters.
```
Comment (serpent5) [inline aspnetcore/web-api/advanced/custom-formatters.md:24]:
```suggestion
Use a custom formatter to add support for a content type that isn't handled by the bult-in formatters.
```

File: aspnetcore/web-api/advanced/custom-formatters.md [hunk 2]
```diff
@@ -9,112 +9,135 @@ uid: web-api/advanced/custom-formatters
...
 
-For example, if some of the clients for your web API can handle the [Protobuf](https://github.com/google/protobuf) format, you might want to use Protobuf with those clients because it's more efficient. Or you might want your web API to send contact names and addresses in [vCard](https://wikipedia.org/wiki/VCard) format, a commonly used format for exchanging contact data. The sample app provided with this article implements a simple vCard formatter.
+For example, if some Web API clients support the [Protobuf](https://github.com/google/protobuf) format, consider using [Protobuf](xref:grpc/dotnet-grpc) because it's more efficient. The sample app provided with this article implements a basic [vCard](https://wikipedia.org/wiki/VCard) formatter. `vCard` is a commonly used format for exchanging contact data.
```
Comment (serpent5) [inline aspnetcore/web-api/advanced/custom-formatters.md:27]:
```suggestion
```

I don't think discussing content-negotiation is appropriate here.

File: aspnetcore/web-api/advanced/custom-formatters.md [hunk 3]
```diff
@@ -9,112 +9,135 @@ uid: web-api/advanced/custom-formatters
...
 
 ## Overview of how to use a custom formatter
 
-Here are the steps to create and use a custom formatter:
+The following steps outline creating a custom formatter:
```
Comment (serpent5) [inline aspnetcore/web-api/advanced/custom-formatters.md:30]:
```suggestion
To create a custom formatter:
```

File: aspnetcore/web-api/advanced/custom-formatters.md [hunk 4]
```diff
@@ -9,112 +9,135 @@ uid: web-api/advanced/custom-formatters
...
 
-* Create an output formatter class if you want to serialize data to send to the client.
-* Create an input formatter class if you want to deserialize data received from the client.
-* Add instances of your formatters to the `InputFormatters` and `OutputFormatters` collections in [MvcOptions](/dotnet/api/microsoft.aspnetcore.mvc.mvcoptions).
-
-The following sections provide guidance and code examples for each of these steps.
+* Create an output formatter class to serialize data sent to the client.
+* Create an input formatter class to deserialize data received from the client.
+* Add instances of formatters to the `InputFormatters` and `OutputFormatters` collections in [MvcOptions](/dotnet/api/microsoft.aspnetcore.mvc.mvcoptions).
```
Comment (serpent5) [inline aspnetcore/web-api/advanced/custom-formatters.md:34]:
```suggestion
* For serializing data sent to the client, create an output formatter class.
* For deserialzing data received from the client, create an input formatter class.
* Add instances of formatter classes to the `InputFormatters` and `OutputFormatters` collections in [MvcOptions](/dotnet/api/microsoft.aspnetcore.mvc.mvcoptions).
```

File: aspnetcore/web-api/advanced/custom-formatters.md [hunk 5]
```diff
@@ -9,112 +9,135 @@ uid: web-api/advanced/custom-formatters
...
 
 To create a formatter:
 
-* Derive the class from the appropriate base class.
+* Derive the class from the appropriate base class. The sample app derives from <xref:Microsoft.AspNetCore.Mvc.Formatters.TextOutputFormatter> and <xref:Microsoft.AspNetCore.Mvc.Formatters.TextInputFormatter>.
 * Specify valid media types and encodings in the constructor.
-* Override `CanReadType`/`CanWriteType` methods
-* Override `ReadRequestBodyAsync`/`WriteResponseBodyAsync` methods
+* Override <xref:Microsoft.AspNetCore.Mvc.Formatters.InputFormatter.CanReadType%2A> and <xref:Microsoft.AspNetCore.Mvc.Formatters.OutputFormatter.CanWriteType%2A> methods
+* Override <xref:Microsoft.AspNetCore.Mvc.Formatters.InputFormatter.ReadRequestBodyAsync%2A> and `WriteResponseBodyAsync` methods
```
Comment (serpent5) [inline aspnetcore/web-api/advanced/custom-formatters.md:43]:
```suggestion
* Override the <xref:Microsoft.AspNetCore.Mvc.Formatters.InputFormatter.CanReadType%2A> and <xref:Microsoft.AspNetCore.Mvc.Formatters.OutputFormatter.CanWriteType%2A> methods.
* Override the <xref:Microsoft.AspNetCore.Mvc.Formatters.InputFormatter.ReadRequestBodyAsync%2A> and `WriteResponseBodyAsync` methods.
```

File: aspnetcore/web-api/advanced/custom-formatters.md [hunk 6]
```diff
@@ -9,112 +9,135 @@ uid: web-api/advanced/custom-formatters
...
+
+The follow code shows the completed `VcardOutputFormatter` class from the [sample](https://github.com/dotnet/AspNetCore.Docs/tree/master/aspnetcore/web-api/advanced/custom-formatters/3.1sample):
```
Comment (serpent5) [inline aspnetcore/web-api/advanced/custom-formatters.md:45]:
```suggestion
The following code shows the `VcardOutputFormatter` class from the [sample](https://github.com/dotnet/AspNetCore.Docs/tree/master/aspnetcore/web-api/advanced/custom-formatters/3.1sample):
```

File: aspnetcore/web-api/advanced/custom-formatters.md [hunk 7]
```diff
@@ -9,112 +9,135 @@ uid: web-api/advanced/custom-formatters
...
-[!code-csharp[](custom-formatters/sample/Formatters/VcardOutputFormatter.cs?name=ctor&highlight=3,5-6)]
-
-For an input formatter example, see the [sample app](https://github.com/dotnet/AspNetCore.Docs/tree/master/aspnetcore/web-api/advanced/custom-formatters/sample).
-
-> [!NOTE]
-> You can't do constructor dependency injection in a formatter class. For example, you can't get a logger by adding a logger parameter to the constructor. To access services, you have to use the context object that gets passed in to your methods. A code example [below](#read-write) shows how to do this.
+[!code-csharp[](custom-formatters/3.1sample/Formatters/VcardOutputFormatter.cs?name=ctor)]
 
-### Override CanReadType/CanWriteType
+Constructor dependency injection can ***not*** be done in a formatter class. For example, the logger cannot be added as logger parameter to the constructor. To access services, use the context object that gets passed in to the methods. A code example in this doc and the [download code](https://github.com/dotnet/AspNetCore.Docs/tree/master/aspnetcore/web-api/advanced/custom-formatters/sample) show how to do this.
```
Comment (serpent5) [inline aspnetcore/web-api/advanced/custom-formatters.md:63]:
```suggestion
A formatter class can **not** use constructor injection for its dependencies. For example, `ILogger<VcardOutputFormatter>` cannot be added as a parameter to the constructor. To access services, use the context object that gets passed in to the methods. A code example in this article and the [sample](https://github.com/dotnet/AspNetCore.Docs/tree/master/aspnetcore/web-api/advanced/custom-formatters/3.1sample) show how to do this.
```

File: aspnetcore/web-api/advanced/custom-formatters.md [hunk 8]
```diff
@@ -9,112 +9,135 @@ uid: web-api/advanced/custom-formatters
...
 * There are derived classes which might be returned at runtime.
-* You need to know at runtime which derived class was returned by the action.
+* The derived class returned by the action must be known at runtime.
 
-For example, suppose your action method signature returns a `Person` type, but it may return a `Student` or `Instructor` type that derives from `Person`. If you want your formatter to handle only `Student` objects, check the type of [Object](/dotnet/api/microsoft.aspnetcore.mvc.formatters.outputformattercanwritecontext.object#Microsoft_AspNetCore_Mvc_Formatters_OutputFormatterCanWriteContext_Object) in the context object provided to the `CanWriteResult` method. Note that it's not necessary to use `CanWriteResult` when the action method returns `IActionResult`; in that case, the `CanWriteType` method receives the runtime type.
+For example, suppose the action method:
 
-<a id="read-write"></a>
+* Signature returns a `Person` type.
+* It can return a `Student` or `Instructor` type that derives from `Person`.
```
Comment (serpent5) [inline aspnetcore/web-api/advanced/custom-formatters.md:82]:
```suggestion
* Can return a `Student` or `Instructor` type that derives from `Person`. 
```

File: aspnetcore/web-api/advanced/custom-formatters.md [hunk 9]
```diff
@@ -9,112 +9,135 @@ uid: web-api/advanced/custom-formatters
...
+For example, suppose the action method:
 
-<a id="read-write"></a>
+* Signature returns a `Person` type.
+* It can return a `Student` or `Instructor` type that derives from `Person`. 
+
+For the formatter to handle only `Student` objects, check the type of [Object](/dotnet/api/microsoft.aspnetcore.mvc.formatters.outputformattercanwritecontext.object#Microsoft_AspNetCore_Mvc_Formatters_OutputFormatterCanWriteContext_Object) in the context object provided to the `CanWriteResult` method. When the action method returns `IActionResult`:
 
-### Override ReadRequestBodyAsync/WriteResponseBodyAsync
+*  It's not necessary to use `CanWriteResult`.
```
Comment (serpent5) [inline aspnetcore/web-api/advanced/custom-formatters.md:86]:
```suggestion
* It's not necessary to use `CanWriteResult`.
```

File: aspnetcore/web-api/advanced/custom-formatters.md [hunk 10]
```diff
@@ -9,112 +9,135 @@ uid: web-api/advanced/custom-formatters
...
+* The `CanWriteType` method receives the runtime type.
 
-You do the actual work of deserializing or serializing in `ReadRequestBodyAsync` or `WriteResponseBodyAsync`. The highlighted lines in the following example show how to get services from the dependency injection container (you can't get them from constructor parameters).
+<a id="read-write"></a>
 
-[!code-csharp[](custom-formatters/sample/Formatters/VcardOutputFormatter.cs?name=writeresponse&highlight=3-4)]
+### Override ReadRequestBodyAsync and WriteResponseBodyAsync
 
-For an input formatter example, see the [sample app](https://github.com/dotnet/AspNetCore.Docs/tree/master/aspnetcore/web-api/advanced/custom-formatters/sample).
+Deserializing or serializing is performed in `ReadRequestBodyAsync` or `WriteResponseBodyAsync`. The following example shows how to get services from the dependency injection container. Services can't be obtained from constructor parameters.
```
Comment (serpent5) [inline aspnetcore/web-api/advanced/custom-formatters.md:93]:
```suggestion
Deserialization or serialization is performed in `ReadRequestBodyAsync` or `WriteResponseBodyAsync`. The following example shows how to get services from the dependency injection container. Services can't be obtained from constructor parameters.
```

File: aspnetcore/web-api/advanced/custom-formatters.md [hunk 11]
```diff
@@ -9,112 +9,135 @@ uid: web-api/advanced/custom-formatters
...
-[!code-csharp[](custom-formatters/sample/Startup.cs?name=mvcoptions&highlight=3-4)]
+::: moniker range=">= aspnetcore-3.0"
+
+[!code-csharp[](custom-formatters/3.1sample/Startup.cs?name=mvcoptions)]
+
+::: moniker-end
+
+::: moniker range="< aspnetcore-3.0"
+
+[!code-csharp[](custom-formatters/3.1sample/Startup.cs?name=mvcoptions&highlight=3-4)]
```
Comment (serpent5) [inline aspnetcore/web-api/advanced/custom-formatters.md:109]:
```suggestion
[!code-csharp[](custom-formatters/sample/Startup.cs?name=mvcoptions&highlight=3-4)]
```

File: aspnetcore/web-api/advanced/custom-formatters.md [hunk 12]
```diff
@@ -9,112 +9,135 @@ uid: web-api/advanced/custom-formatters
...
+::: moniker-end
 
 Formatters are evaluated in the order you insert them. The first one takes precedence.
 
-## Next steps
+## The completed `VcardInputFormatter` class
+
+The follow code shows the completed `VcardInputFormatter` class from the [sample](https://github.com/dotnet/AspNetCore.Docs/tree/master/aspnetcore/web-api/advanced/custom-formatters/3.1sample):
+
+[!code-csharp[](custom-formatters/3.1sample/Formatters/VcardInputFormatter.cs?name=snippet)]
```
Comment (serpent5) [inline aspnetcore/web-api/advanced/custom-formatters.md:119]: I don't think this section is needed, as the complete class is shown further up, in "How to create a custom formatter class".

File: aspnetcore/web-api/advanced/custom-formatters.md [hunk 13]
```diff
@@ -9,112 +9,135 @@ uid: web-api/advanced/custom-formatters
...
 
-* [Sample app for this doc](https://github.com/dotnet/AspNetCore.Docs/tree/master/aspnetcore/web-api/advanced/custom-formatters/sample), which implements simple vCard input and output formatters. The apps reads and writes vCards that look like the following example:
+## Test the app
+
+[Run the sample app for this doc](https://github.com/dotnet/AspNetCore.Docs/tree/master/aspnetcore/web-api/advanced/custom-formatters/sample), which implements basic vCard input and output formatters. The app reads and writes vCards similar to the following:
```
Comment (serpent5) [inline aspnetcore/web-api/advanced/custom-formatters.md:123]:
```suggestion
[Run the sample app for this article](https://github.com/dotnet/AspNetCore.Docs/tree/master/aspnetcore/web-api/advanced/custom-formatters/sample), which implements basic vCard input and output formatters. The app reads and writes vCards similar to the following:
```

File: aspnetcore/web-api/advanced/custom-formatters.md [hunk 14]
```diff
@@ -9,112 +9,135 @@ uid: web-api/advanced/custom-formatters
...
+To add a vCard to the in-memory collection of contacts:
+
+* Send a `Post` request to `/api/contacts` with a tool like Postman.
+* Set the `Content-Type` header to `text/vcard`.
+* Set `vCard` text in the body, formatted like the preceding example.
+
+## Additional resources
 
-To add a vCard to the in-memory collection of contacts, send a Post request to the same URL, with Content-Type header "text/vcard" and with vCard text in the body, formatted like the example above.
+<xref:grpc/dotnet-grpc>
```
Comment (serpent5) [inline aspnetcore/web-api/advanced/custom-formatters.md:143]:
```suggestion
* <xref:web-api/advanced/formatting>
* <xref:grpc/dotnet-grpc>
```

**Comment by serpent5** [comment]
> @serpent5 the 2.2 and the new 3.1 version work fine with GET, but on POST the model is null.

@Rick-Anderson Did you resolve this issue with POST?

File: aspnetcore/web-api/advanced/custom-formatters.md (hunk 12 above)
Comment (Rick-Anderson) [reply aspnetcore/web-api/advanced/custom-formatters.md:117]: That's the `VcardOutputFormatter ` class, this is the input class.

**Comment by Rick-Anderson** [comment]
> There's a few code suggestions, too, but it'll be easier if I can do those in a follow-up PR. Those suggestions will be more along the lines of cleaning up a little, but nothing major.

I think we should use that approach when possible.

File: aspnetcore/web-api/advanced/custom-formatters.md (hunk 12 above)
Comment (serpent5) [reply aspnetcore/web-api/advanced/custom-formatters.md:117]: Ha, yes, of course! Sorry, I totally missed that. One more suggestion for that part coming up then cos of course it's staying. 😅

File: aspnetcore/web-api/advanced/custom-formatters.md [hunk 15]
```diff
@@ -9,112 +9,134 @@ uid: web-api/advanced/custom-formatters
...
 [!code-csharp[](custom-formatters/sample/Startup.cs?name=mvcoptions&highlight=3-4)]
 
+::: moniker-end
+
 Formatters are evaluated in the order you insert them. The first one takes precedence.
 
-## Next steps
+## The completed `VcardInputFormatter` class
+
+The follow code shows the completed `VcardInputFormatter` class from the [sample](https://github.com/dotnet/AspNetCore.Docs/tree/master/aspnetcore/web-api/advanced/custom-formatters/3.1sample):
```
Comment (serpent5) [inline aspnetcore/web-api/advanced/custom-formatters.md:115]:
```suggestion
The following code shows the `VcardInputFormatter` class from the [sample](https://github.com/dotnet/AspNetCore.Docs/tree/master/aspnetcore/web-api/advanced/custom-formatters/3.1sample):
```

**Comment by serpent5** [comment]
@Rick-Anderson Do you prefer "doc", "topic", or "article"? I've been going with "topic" but I went with "article" in the suggestions since I just realised that the inline ToC thing always says "In this _article_".

**Comment by Rick-Anderson** [comment]
> Do you prefer "doc", "topic", or "article"? 
Never doc, perhaps document. `article` seems like a good choice but `topic` is probably OK. Consistency is important.

**Pull request merged**